### PR TITLE
feat: use just two SQL calls to back the incidents API

### PR DESCRIPTION
**Pull request by srabraham**
Currently a call to listIncidents (the API that backs the incidents page) results in over 5*n sequential DB lookups, where n is the number of incidents in the relevant event. Given that we build to over 1000 incidents per event, that starts to be quite a lot of queries for each incidents page load. My suspicion is that it'd be much faster to issue just a few big queries instead, and that's what this change does.

On staging, before this change, a listIncidents for 2023 takes about 10 seconds and issues over 6000 DB selects. I'm hopeful that this change will bring that duration down significantly. https://ranger-ims-staging.burningman.org/ims/api/events/2023/incidents/

I've confirmed locally for MySQL and for SQLite that this change doesn't alter the API result for my bunch of test data. The DataStoreIncidentTests pass on this change too, so that's pretty cool.

https://github.com/burningmantech/ranger-ims-server/issues/1324

### CODE REVIEW

File: src/ims/store/_db.py
Comment (wsanchez) [inline src/ims/store/_db.py:594]: Quote from the future (via GitHub Time Machine™):
> But @wsanchez, you are such a liar! Right here is `incidents_reportEntries`, which is your fault!

This is true. The `_` indicates a subcategory of some sort. In this case, it's `<tableName>_<queryName>`.

File: src/ims/store/mysql/_queries.py
Comment (wsanchez) [inline src/ims/store/mysql/_queries.py:232]: `json_arrayagg` is too cool for me, so I'll just assume this is sane. 🙃

**Review by wsanchez — CHANGES_REQUESTED** [review]
If you could match the code style on the variable names, that would be swell.
Looks good otherwise.

File: src/ims/store/mysql/_queries.py
Comment (srabraham) [reply src/ims/store/mysql/_queries.py:232]: haha, yeah, I just learned about it myself. It's relatively new. It takes a bunch of things and turns them into a JSON array (i.e. a string that can be parsed as a JSON array)

File: src/ims/store/_db.py
Comment (wsanchez) [inline src/ims/store/_db.py:767]: Did you confirm this is better?

FWIW, if you add `--log-level=debug` to server startup and look for messages in the namespace `ims.store.mysql.*` (or `ims.store.sqlite.*`), you'll see all of the DB transactions are logged, eg:

```console
2024-10-31T11:19:49-0700 [ims.store.mysql._store.Cursor#debug] EXECUTE: 
	select VERSION from SCHEMA_INFO
	 <- {}
2024-10-31T11:19:49-0700 [ims.store._db.DatabaseManager#debug] No upgrade required for schema version 6
2024-10-31T11:19:49-0700 [ims.store.mysql._store.DataStore#info] Validating data store...
2024-10-31T11:19:50-0700 [ims.store.mysql._store.Cursor#debug] EXECUTE: 
	select AUTHOR, TEXT, CREATED, GENERATED from REPORT_ENTRY
	where
	    ID not in (select REPORT_ENTRY from INCIDENT__REPORT_ENTRY) and
	    ID not in (select REPORT_ENTRY from INCIDENT_REPORT__REPORT_ENTRY)
	 <- {}
```

File: src/ims/store/_db.py
Comment (srabraham) [reply src/ims/store/_db.py:594]: Good to know!

File: src/ims/store/_db.py
Comment (srabraham) [reply src/ims/store/_db.py]: ahh, ok, I'd wondered about that...it's been really weird seeing camelcase in Python. I agree consistency is important. Done

File: src/ims/store/_db.py
Comment (srabraham) [reply src/ims/store/_db.py:766]: Yeah, I have been watching those logs locally, with a small bunch of incidents and incident entries
I can only run IMS locally and see the logs locally as things are, so my intention is to merge the PR to see the impact on API duration in staging

**Comment by srabraham** [comment]
I've addressed the review comments. The "change requested" review status is blocking me from merging myself

**Comment by srabraham** [comment]
> If you could match the code style on the variable names, that would be swell. Looks good otherwise.

Looks like this is an approval, given that I've addressed all the comments. I'm going to dismiss the "changes requested", which otherwise blocks merging

**Pull request merged**
